core/src/main/java/io/grpc/internal/ServerImpl.java

```diff
@@ -509,20 +511,37 @@ public void cancelled(Context context) {
         ServerMethodDefinition<ReqT, RespT> methodDef, Metadata headers,
         Context.CancellableContext context, StatsTraceContext statsTraceCtx) {
       // TODO(ejona86): should we update fullMethodName to have the canonical path of the method?
-      ServerCallImpl<ReqT, RespT> call = new ServerCallImpl<ReqT, RespT>(
-          stream, methodDef.getMethodDescriptor(), headers, context,
-          decompressorRegistry, compressorRegistry);
-      ServerCallHandler<ReqT, RespT> callHandler = methodDef.getServerCallHandler();
       statsTraceCtx.serverCallStarted(
           new ServerCallInfoImpl<ReqT, RespT>(
-              methodDef.getMethodDescriptor(),
-              call.getAttributes(),
-              call.getAuthority()));
-
+              methodDef.getMethodDescriptor(), // notify with original method descriptor
+              stream.getAttributes(),
+              stream.getAuthority()));
+      ServerCallHandler<ReqT, RespT> handler = methodDef.getServerCallHandler();
       for (ServerInterceptor interceptor : interceptors) {
-        callHandler = InternalServerInterceptors.interceptCallHandler(interceptor, callHandler);
+        handler = InternalServerInterceptors.interceptCallHandler(interceptor, handler);
       }
-      ServerCall.Listener<ReqT> listener = callHandler.startCall(call, headers);
+      ServerMethodDefinition<ReqT, RespT> interceptedDef = methodDef.withServerCallHandler(handler);
+      ServerMethodDefinition<?, ?> wMethodDef = binlogProvider == null
+          ? interceptedDef : binlogProvider.wrapMethodDefinition(interceptedDef);
+      return startWrappedCall(fullMethodName, wMethodDef, stream, headers, context);
+    }
+
+    private <WReqT, WRespT> ServerStreamListener startWrappedCall(
```

Comment:
Bazel complains unless this method is bound to differently named types:
```
core/src/main/java/io/grpc/internal/ServerImpl.java:525: error: method startCall0 in class ServerImpl.ServerTransportListenerImpl cannot be applied to given types;
      return startCall0(
             ^
  required: String,ServerMethodDefinition<ReqT#1,RespT#1>,ServerStream,Metadata,CancellableContext
  found: String,binlogProv[...]dDef),ServerStream,Metadata,CancellableContext
  reason: inference variable ReqT#1 has incompatible equality constraints CAP#1,ReqT#2
  where ReqT#1,RespT#1,ReqT#2,RespT#2 are type-variables:
    ReqT#1 extends Object declared in method <ReqT#1,RespT#1>startCall0(String,ServerMethodDefinition<ReqT#1,RespT#1>,ServerStream,Metadata,CancellableContext)
    RespT#1 extends Object declared in method <ReqT#1,RespT#1>startCall0(String,ServerMethodDefinition<ReqT#1,RespT#1>,ServerStream,Metadata,CancellableContext)
    ReqT#2 extends Object declared in method <ReqT#2,RespT#2>startCall(ServerStream,String,ServerMethodDefinition<ReqT#2,RespT#2>,Metadata,CancellableContext,StatsTraceContext)
    RespT#2 extends Object declared in method <ReqT#2,RespT#2>startCall(ServerStream,String,ServerMethodDefinition<ReqT#2,RespT#2>,Metadata,CancellableContext,StatsTraceContext)
  where CAP#1 is a fresh type-variable:
    CAP#1 extends Object from capture of ?
```

Reply:
That error message... what is this? C++?

Guess it doesn't matter much. Choosing a different name isn't too annoying.